README.md
Copyright 2013 NPR.  All rights reserved.  No part of these materials may be reproduced, modified, stored in a retrieval system, or retransmitted, in any form or by any means, electronic, mechanical or otherwise, without prior written permission from NPR.

(Want to use this code? Send an email to nprapps@npr.org!)

Tiny Desk Concerts Roku app
===========================

Setup
-----

Put your Roku into [Dev Mode](http://sdkdocs.roku.com/display/RokuSDKv48/Developer+Guide#DeveloperGuide-71EnablingDevelopmentModeonyourbox).

Your Roku will tell you its IP address and have you set a developer password. Set these in your `.bash_profile`:

```
export ROKU_IP="10.0.1.2"
export ROKU_PASSWORD="mypassword"
```

Install the Python requirements:

```
$ mkvirtualenv roku-tinydesk
$ pip install requirements.txt
```

Generate JSON feed
------------------

**This step is not necessary to run the app.**

To regenerate the API feed you will need a superuser NPR API key and our AWS keys. Set the API key in your `.bash_profile` like so:

```
export NPR_API_KEY="SUPERUSER_KEY_HERE"
```

Install `s3cmd` and configure it with the `nprapps` AWS keys:

```
$ s3cmd --configure
```

To generate (and deploy) the Tiny Desk API feed:

```
$ workon roku-tinydesk 
$ ./fetch_feed.py
```

Package & Deploy
----------------

To package up the code and deploy it to your Roku:

```
$ ./install.sh
```

Enjoy!

Distribute the app
------------------

**Note:** If this process isn't completed with Chris' Roku keys, users of the app will lose their registry data. The way around this is to "rekey" each developer's Roku with the same RSA key. Read more here:

http://sdkdocs.roku.com/display/RokuSDKv48/Channel+Packaging+And+Publishing#ChannelPackagingAndPublishing-30PackagingApplications

**Note part 2:** This is probably irrelevant anyway, because you would need Chris' Roku credentials to access the upload page. Ergo, these docs are really just for him.

Distributing the app to Roku requires jumping through additional hoops. First, you'll need to generate crypotgraphic keys on your Roku. Do this like so:

```
$ telnet $ROKU_IP 8080
> genkey
```

Your password will be print out: **save it**. Exit the telnet session by typing `quit`.

Add this new password to your `~/.bash_profile`:

```
export ROKU_GENKEY_PASSWORD="YOUR_PASSWORD_HERE"
```

Now to generate a shippable zip, simply run:

```
$ ./ship.sh 1.0
```

You must supply a version number as the first argument.

Now take your new package and upload it at: https://owner.roku.com/Developer/Apps/Packages/23962

Debugging
---------

To connect to your Roku for debugging:

```
$ telnet $ROKU_IP 8085
```

Google Analytics
----------------

The following events are tracked in Google Analytics:

|Category|Action|Label|Value|Custom 1|Custom 2|
|--------|------|-----|-----|--------|--------|
|Tiny Desk|Startup|||`rokuModel`|`rokuFirmware`|
|Tiny Desk|Shutdown||`sessionDuration`|`numWatched`|`numFinished`|
|Tiny Desk|Start|`title`||`fromList`|`searchTerm`|
|Tiny Desk|Continue|`title`||`fromList`|`searchTerm`|
|Tiny Desk|Finish|`title`||||
|Tiny Desk|Stop|`title`|`playtime`|`stoppedAtPct`||
|Tiny Desk|Search|`term`||||
|Tiny Desk|Feed Unavailable|||||

**Notes**:

* The *Shutdown* action isn't recorded until the next time the application is run, because there is no reliable shutdown event in a Roku app. As a result, `sessionDuration` will always be approximate. 
* The *Stop* action is reported alongside the *Finish* action when a video is completed, for purposes of tracking playtimes.
* There is no concept of a *ping* in this event model, so users will appear inactive on the Google Analytics dashboard while in the middle of a long video or sitting on a menu. Their `sessionDuration` recorded in the *Shutdown* event will correctly include the this time.
* The *Feed Unavailable* action occurs if the `feed.json` file can not be downloaded from S3 after several attempts. (Do to an error or poor connectivity.)

Code conventions
----------------

Because Roku dev lacks strong (or any) code conventions, I'm defining my own.

**Syntax**

* Constants: `UPPERCASE`
* Variables, public members and public functions: `camelCase`
* Objects/Constructors: `TitleCase`
* Public member functions: `GridScreen.refreshLists = GridScreen_refreshLists`
* Private member functions `GridScreen._initLists = _GridScreen_initLists`
* Private member vars: `_underscorePrefixed`
* Keywords: lowercase, e.g. `function` and `end function`

All rules for function names also apply when calling case-insensitive built-in functions, e.g. `createObject()` and `count()`.

**Code organization**

Leave a line of whitespace after `function` and before `end function`

Never use `sub`. Always use `function`.

**Classes**

In class constructors define the class object as `this`. In member functions redefine the class object as `this = m`. *Never* use `m` to refer to the global namespace.

**Comments**

Comments describing a functions purpose should *precede* the function defintion.

Modules should begin with a block comment explaning their purpose:

```
'
' Explainer goes here
'

function main()
    blah blah
```


CHANGELOG
1.1
---

* Slimmed down package size

1.0
---

* Initial beta release.


fabfile.py
#!/usr/bin/env python

from fabric.api import *

import app_config

"""
Base configuration
"""
env.deployed_name = app_config.PROJECT_SLUG
env.repo_name = app_config.REPOSITORY_NAME

env.repo_url = 'git@github.com:nprapps/%(repo_name)s.git' % env
env.alt_repo_url = None  # 'git@bitbucket.org:nprapps/%(repo_name)s.git' % env
env.user = 'ubuntu'
env.python = 'python2.7'
env.path = '/home/%(user)s/apps/%(deployed_name)s' % env
env.repo_path = '%(path)s/repository' % env
env.virtualenv_path = '%(path)s/virtualenv' % env
env.forward_agent = True

"""
Environments
"""
def production():
    env.settings = 'production'
    env.hosts = app_config.PRODUCTION_SERVERS

def staging():
    env.settings = 'staging'
    env.hosts = app_config.STAGING_SERVERS

"""
Branches
"""
def stable():
    """
    Work on stable branch.
    """
    env.branch = 'stable'

def master():
    """
    Work on development branch.
    """
    env.branch = 'master'

def branch(branch_name):
    """
    Work on any specified branch.
    """
    env.branch = branch_name

"""
Setup
"""
def setup():
    """
    Setup servers for deployment.
    """
    require('settings', provided_by=[production, staging])
    require('branch', provided_by=[stable, master, branch])

    setup_directories()
    create_log_file()
    setup_virtualenv()
    clone_repo()
    checkout_latest()
    install_requirements()

def setup_directories():
    """
    Create server directories.
    """
    require('settings', provided_by=[production, staging])

    run('mkdir -p %(path)s' % env)

def setup_log_file():
    sudo('touch /var/log/%(deployed_name)s.log' % env)
    sudo('chown %(user)s /var/log/%(deployed_name)s.log' % env)

def setup_virtualenv():
    """
    Setup a server virtualenv.
    """
    require('settings', provided_by=[production, staging])

    run('virtualenv -p %(python)s --no-site-packages %(virtualenv_path)s' % env)
    run('source %(virtualenv_path)s/bin/activate' % env)

def clone_repo():
    """
    Clone the source repository.
    """
    require('settings', provided_by=[production, staging])

    run('git clone %(repo_url)s %(repo_path)s' % env)

    if env.get('alt_repo_url', None):
        run('git remote add bitbucket %(alt_repo_url)s' % env)

def checkout_latest(remote='origin'):
    """
    Checkout the latest source.
    """
    require('settings', provided_by=[production, staging])
    require('branch', provided_by=[stable, master, branch])

    env.remote = remote

    run('cd %(repo_path)s; git fetch %(remote)s' % env)
    run('cd %(repo_path)s; git checkout %(branch)s; git pull %(remote)s %(branch)s' % env)

def install_requirements():
    """
    Install the latest requirements.
    """
    require('settings', provided_by=[production, staging])

    run('%(virtualenv_path)s/bin/pip install -U -r %(repo_path)s/requirements.txt' % env)

def install_crontab():
    """
    Install cron jobs script into cron.d.
    """
    require('settings', provided_by=[production, staging])

    sudo('cp %(repo_path)s/crontab /etc/cron.d/%(deployed_name)s' % env)

def uninstall_crontab():
    """
    Remove a previously install cron jobs script from cron.d
    """
    require('settings', provided_by=[production, staging])

    sudo('rm /etc/cron.d/%(deployed_name)s' % env)

"""
Deployment
"""
def deploy(remote='origin'):
    """
    Deploy the latest app to S3 and, if configured, to our servers.
    """
    require('settings', provided_by=[production, staging])

    require('branch', provided_by=[stable, master, branch])

    if (env.settings == 'production' and env.branch != 'stable'):
        _confirm("You are trying to deploy the '%(branch)s' branch to production.\nYou should really only deploy a stable branch.\nDo you know what you're doing?" % env)

    checkout_latest(remote)
    install_crontab()

"""
Destruction
"""
def _confirm(message):
    answer = prompt(message, default="Not at all")

    if answer.lower() not in ('y', 'yes', 'buzz off', 'screw you'):
        exit()

def shiva_the_destroyer():
    """
    Deletes the app from the servers 
    """
    require('settings', provided_by=[production, staging])

    _confirm("You are about to destroy everything deployed to %(settings)s for this project.\nDo you know what you're doing?" % env)

    with settings(warn_only=True):
        run('rm -rf %(path)s' % env)
        uninstall_crontab()



app_config.py
#!/usr/bin/env python

"""
Project-wide application configuration.

DO NOT STORE SECRETS, PASSWORDS, ETC. IN THIS FILE.
They will be exposed to users. Use environment variables instead.
"""

import os

PROJECT_NAME = 'Tiny Desk'
PROJECT_SLUG = 'roku-tinydesk'
REPOSITORY_NAME = 'roku-tinydesk'

PRODUCTION_S3_BUCKETS = ['apps.npr.org']
PRODUCTION_SERVERS = ['cron.nprapps.org']

STAGING_S3_BUCKETS = ['stage-apps.npr.org']
STAGING_SERVERS = ['cron-staging.nprapps.org']

S3_BUCKETS = []
SERVERS = []
DEBUG = True

def configure_targets(deployment_target):
    """
    Configure deployment targets. Abstracted so this can be
    overriden for rendering before deployment.
    """
    global S3_BUCKETS
    global SERVERS
    global DEBUG

    if deployment_target == 'production':
        S3_BUCKETS = PRODUCTION_S3_BUCKETS
        SERVERS = PRODUCTION_SERVERS
        DEBUG = False
    else:
        S3_BUCKETS = STAGING_S3_BUCKETS
        SERVERS = STAGING_SERVERS
        DEBUG = True

DEPLOYMENT_TARGET = os.environ.get('DEPLOYMENT_TARGET', None)

configure_targets(DEPLOYMENT_TARGET)


fetch_feed.py
#!/usr/bin/env python

from HTMLParser import HTMLParser
import gzip
import json
import os
import shlex
import string
import subprocess
import unicodedata

from dateutil.parser import parse
import requests

BITRATES = [200000, 500000, 1000000, 2000000]
UPLOAD_CMD = 's3cmd -P --add-header=Cache-Control:max-age=5 --add-header=Content-encoding:gzip --guess-mime-type sync feed.json s3://apps.npr.org/roku-tinydesk/'

class MLStripper(HTMLParser):
    def __init__(self):
        self.reset()
        self.fed = []

    def handle_data(self, d):
        self.fed.append(d)

    def get_data(self):
        return ''.join(self.fed)

def strip_tags(html):
    s = MLStripper()
    s.feed(html)
    return s.get_data()

def main():
    output = []
    page = 0

    missing_mp4s = ''

    while True: 
        print 'Fetching page %i' % page
        url = 'http://api.npr.org/query?id=92071316&apiKey=%s&output=json&startNum=%i&numResults=40' % (os.environ['NPR_API_KEY'], page * 40)
        print url
        response = requests.get(url)

        data = response.json() 

        if 'message' in data and data['message'][0]['id'] == '401':
            print data['message']
            break

        for story in data['list']['story']:
            title = story['title']['$text'].replace(': Tiny Desk Concert', '')
            print title

            item = {
                'id': story['id'],
                'title': title,
                'searchTitle': searchify_title(title), 
                'sortTitle': sortify_title(title), 
                'titleSeason': 'Tiny Desk Concerts',
                'description': strip_tags(story['miniTeaser']['$text']),
                'sdPosterUrl': None,
                'hdPosterUrl': None,
                'length': int(story['multimedia'][0]['duration']['$text']),
                'releaseDate': None,
                'streamFormat': 'mp4',
                'streamBitrates': [],
                'streamUrls': [],
                'streamQualities': [],
                'isHD': True,
                'hdBranded': True
            }

            alt_image_url = story['multimedia'][0]['altImageUrl']['$text']

            # Yes, I know this is a hack, but we don't want to urlencode the protocol or domain
            item['sdPosterUrl'] = alt_image_url.replace(' ', '%20') + '?s=2'
            item['hdPosterUrl'] = alt_image_url.replace(' ', '%20') + '?s=3'

            # Formatted as: "Mon, 11 Mar 2013 14:03:00 -0400"
            pub_date = story['pubDate']['$text']
            dt = parse(pub_date)

            item['releaseDate'] = dt.strftime('%B %d, %Y') 

            if 'mp4' not in story['multimedia'][0]['format']:
                print '--> No mp4 video, skipping!'

                missing_mp4s += '%s, %s\n' % (item['id'], item['title'])
                
                continue

            video_url = story['multimedia'][0]['format']['mp4']['$text']
            
            for bitrate in BITRATES:
                item['streamBitrates'].append(int(bitrate / 1024))
                item['streamUrls'].append(video_url.replace('-n', '-n-%i' % bitrate))
                item['streamQualities'].append('SD')

            # The last item is full resolution 
            item['streamUrls'][-1] = video_url
            item['streamQualities'][-1] = 'HD'

            output.append(item)

        page += 1

    print 'Saving %i concerts' % len(output)

    with gzip.open('feed.json', 'wb') as f:
        f.write(json.dumps(output))

    with open('missing_mp4s.txt', 'w') as f:
        f.write(missing_mp4s)

    print 'Deploying to S3'

    subprocess.Popen(shlex.split(UPLOAD_CMD), stdout=subprocess.PIPE, stderr=subprocess.PIPE)

def searchify_title(title):
    return ''.join(x for x in unicodedata.normalize('NFKD', title) if x in string.printable)

def sortify_title(title):
    title = searchify_title(title)

    title = ''.join(x for x in title if x in string.letters + string.digits + ' ').lower()

    if title[0:4] == 'the ':
        title = title[4:]

    return title

if __name__ == '__main__':
    main()


update_feed.py
#!/usr/bin/env python

import csv
from HTMLParser import HTMLParser
import gzip
import json
import logging
import os
import shlex
import string
import subprocess
import unicodedata

from dateutil.parser import parse
import requests

import app_config

BITRATES = [200000, 500000, 1000000, 2000000]
UPLOAD_CMD = 's3cmd -P --add-header=Cache-Control:max-age=5 --add-header=Content-encoding:gzip --guess-mime-type sync feed.json s3://%s/roku-tinydesk/' % app_config.S3_BUCKETS[0]

logging.basicConfig(filename='/var/log/roku-tinydesk.log', level=logging.DEBUG, format='%(asctime)s: %(message)s')

class MLStripper(HTMLParser):
    def __init__(self):
        self.reset()
        self.fed = []

    def handle_data(self, d):
        self.fed.append(d)

    def get_data(self):
        return ''.join(self.fed)

def strip_tags(html):
    s = MLStripper()
    s.feed(html)
    return s.get_data()

def main():
    with open('greatest_hits.txt') as f:
        reader = csv.reader(f)
        greatest_hits = [row[1].strip() for row in reader]

    with open('skip_list.txt') as f:
        reader = csv.reader(f)
        skip_list = [row[1].strip() for row in reader]

    output = []
    page = 0

    missing_mp4s = ''
    zero_length = ''

    while True: 
        logging.info('Fetching page %i' % page)
        url = 'http://api.npr.org/query?id=92071316&apiKey=%s&output=json&startNum=%i&numResults=40' % (os.environ['NPR_API_KEY'], page * 40)
        logging.debug(url)
        response = requests.get(url)

        data = response.json() 

        if 'message' in data and data['message'][0]['id'] == '401':
            break

        for story in data['list']['story']:
            title = story['title']['$text'].replace(': Tiny Desk Concert', '')
            logging.debug(title)

            # Skip any story we've added to the skip list
            if story['id'] in skip_list:
                logging.info('--> In skip list, skipping!')
                continue

            if 'multimedia' not in story or len(story['multimedia']) == 0:
                logging.info('--> No multimedia element, skipping!')
                continue

            item = {
                'id': story['id'],
                'title': title,
                'searchTitle': searchify_title(title), 
                'sortTitle': sortify_title(title), 
                'titleSeason': 'Tiny Desk Concerts',
                'description': strip_tags(story['miniTeaser']['$text']),
                'sdPosterUrl': None,
                'hdPosterUrl': None,
                'length': int(story['multimedia'][0]['duration'].get('$text', 0)),
                'releaseDate': None,
                'streamFormat': 'mp4',
                'streamBitrates': [],
                'streamUrls': [],
                'streamQualities': [],
                'isHD': True,
                'hdBranded': True,
                'greatestHit': None 
            }

            alt_image_url = story['multimedia'][0]['altImageUrl']['$text']

            # Yes, I know this is a hack, but we don't want to urlencode the protocol or domain
            item['sdPosterUrl'] = alt_image_url.replace(' ', '%20') + '?s=2'
            item['hdPosterUrl'] = alt_image_url.replace(' ', '%20') + '?s=3'

            # Formatted as: "Mon, 11 Mar 2013 14:03:00 -0400"
            story_date = story['storyDate']['$text']
            dt = parse(story_date)

            item['tempDate'] = dt
            item['releaseDate'] = dt.strftime('%B ') + dt.strftime('%d, ').lstrip('0') + dt.strftime('%Y') 

            try:
                item['greatestHit'] = greatest_hits.index(story['id'])
            except ValueError:
                pass

            if 'mp4' not in story['multimedia'][0]['format']:
                logging.debug('--> No mp4 video, skipping!')

                missing_mp4s += '%s, %s\n' % (item['id'], item['title'])
                
                continue

            if item['length'] == 0:
                logging.debug('--> Zero length, including anyway!')

                zero_length += '%s, %s\n' % (item['id'], item['title'])


            video_url = story['multimedia'][0]['format']['mp4']['$text']
            
            for bitrate in BITRATES:
                item['streamBitrates'].append(int(bitrate / 1024))
                item['streamUrls'].append(video_url.replace('-n', '-n-%i' % bitrate))
                item['streamQualities'].append('SD')

            # The last item is full resolution 
            item['streamUrls'][-1] = video_url
            item['streamQualities'][-1] = 'HD'

            output.append(item)

        page += 1

    output = sorted(output, key=lambda k: k['tempDate'])
    output.reverse()
    
    for item in output:
        del item['tempDate']

    if len(output) < 200:
        logging.error('Less than 200 concerts found--sanity check failed! Not updating feed.')
        return

    logging.info('Saving %i concerts' % len(output))

    with gzip.open('feed.json', 'wb') as f:
        f.write(json.dumps(output))

    with open('missing_mp4s.txt', 'w') as f:
        f.write(missing_mp4s)

    with open('zero_length.txt', 'w') as f:
        f.write(zero_length)

    logging.info('Deploying to S3')

    logging.debug(UPLOAD_CMD)
    subprocess.Popen(shlex.split(UPLOAD_CMD), stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    logging.info('Done!')

def searchify_title(title):
    return ''.join(x for x in unicodedata.normalize('NFKD', title) if x in string.printable)

def sortify_title(title):
    title = searchify_title(title)

    title = ''.join(x for x in title if x in string.letters + string.digits + ' ').lower()

    if title[0:4] == 'the ':
        title = title[4:]

    return title

if __name__ == '__main__':
    main()
